pkg/nfd-master/nfd-master.go

```diff
@@ -432,6 +436,85 @@ func (m *nfdMaster) SetLabels(c context.Context, r *pb.SetLabelsRequest) (*pb.Se
 			klog.Errorf("failed to advertise labels: %v", err)
 			return &pb.SetLabelsReply{}, err
 		}
+
+		// ensure we delete the old stale taints
+		labelReply, err := m.deleteTaints(c, r)
+		if err != nil {
+			return labelReply, err
+		}
+
+		if m.args.EnableTaints {
+			// Set NodeFeatureRule provided taints
+			if mustTaint {
+				labelReply, err := m.setTaints(c, r)
+				if err != nil {
+					return labelReply, err
+				}
+			}
+		}
```

Comment:
Hmm, we need to think a bit more this taint handling 🧐 It would be nice to handle taints by patching in `updateNodeFeatures()` similarly to other changes we do on the Node object. I just think that patching isn't an option in this case as taints are an array and we cannot unambiguously and reliably specify elements that we want to delete (a classic race: the node object and the list of taints might've changed between we read the object and send the patch).

So, I think we're stuck with updating the object. But with this. two node updates is an overkill (first delete and then add) plus there is a small time window when nfd taints do not exist(!) So I think a separate one function to handle this would be needed. 

1. get node
2. see if taints need to be updated
3. if not, return
4. if an update is needed, update the taint array (delete, modify, add)
5. update annotations
6. update node (only if smth was changed)

That function could have taints as an arg so that the function could be called with empty taints if `m.args.EnableTaints==true`, deleting anything managed by NFD.

Also, doing node updates we re-introduce this race: #122. I think some re-try mechanism (of few tries) would be needed to mitigate that.

Thoughts?

BTW, we also need to delete taints in `prune()`, too

Reply:
Thanks for the feedback @marquiz . I've refactored the code on my latest commit https://github.com/kubernetes-sigs/node-feature-discovery/pull/910/commits/6331eadb96c8bf45e60e62b1f66bed295a49a84e as per suggestion. Please have a look, if looks good I will squash the commit with prior once.